### ISSUE TITLE
Don't write files in place on completion request

### ISSUE THREAD

**Issue by hjz** [ensime/ensime-server#155]
ensime-write-buffer-to-temp-file seems to be missing from ensime.el in 5.0

**Comment by aemoncannon** [comment]
@edani
FYI I'm working on a cleanup of the completion stuff. Mostly just separating auto-complete backend from company backend. Afterward I want to give this a serious look.

**Comment by edani** [comment]
@aemoncannon If you can wait a few days, I made more server-side changes and I need to push client changes to go with them.

**Comment by aemoncannon** [comment]
@edani You mean wait on the re-org?

**Comment by edani** [comment]
@aemoncannon  I meant wait on looking at this issue (#155). I'll have an update on this shortly.

**Comment by aemoncannon** [comment]
@edani
Okey dokey.

FWIW I just ran these benchmarks:

(defun bench-file ()
    (benchmark-run 1000
      (ensime-eval `(swank:typecheck-file ,buffer-file-name))))


(defun bench-contents ()
    (benchmark-run 1000
      (ensime-eval `(swank:typecheck-file ,buffer-file-name
,(ensime-get-buffer-as-string)))))


;; Result of 3 runs each, where each run is: (total-elapsed-time #of-GCs
#total-GC-time)

;; to disk
(2.057487 2 0.14408899999999925)
(2.194499 3 0.2171789999999998)
(2.070298 2 0.1403800000000004)

;; by contents
(11.508381 23 1.6244739999999993)
(11.635637000000001 23 1.584938000000001)
(11.490751 22 1.5078890000000005)

**Comment by aemoncannon** [comment]
This was on Analyzer.scala with ~11k characters.

**Comment by rorygraves** [comment]
Wow, thats quite a difference, I guess this is encoding and decoding the buffer in emacs?

**Comment by aemoncannon** [comment]
I suspect its the string creation yeah, but we can tease out the culprit
with profiler

**Comment by fommil** [comment]
when I finally get round to moving us fully over to S-Express for the protocol, the contents-over-the-wire might be better.

**Comment by rorygraves** [comment]
I'm guessing it's the enacts rather than jvm side.

**Comment by aemoncannon** [comment]
Here's a dump from the profiler (doesn't add up to 100 because of keyboard
handling etc that's not part of the bench). The gist is that it *looks*
like it's spending most of the time waiting on the socket.

   - bench-contents                                            1826  47%
      - let                                                      1826  47%
       - list                                                    1826  47%
        - -                                                      1826  47%
         - let                                                   1826  47%
          - let                                                  1826  47%
           - while                                               1826  47%
            - ensime-eval                                        1826  47%
             - let*                                              1826  47%
              - apply                                            1825  47%
               - catch                                           1825  47%
                - let                                            1825  47%
                 - while                                         1313  33%
                  - accept-process-output                          53   1%
                   - ensime-net-filter                             16   0%
                    - ensime-process-available-input                 16   0

**Comment by aemoncannon** [comment]
A note on scaling: DebugManager.scala is 3 x the size of Analyzer.scala.
Bench-file takes, again, ~2s. Bench-contents takes 16s.

Caveat:  I bet this is very OS and Platform dependent. I'm on OSX with a
solid state drive. Though, for bench-file,  I'm guessing all the reads are
coming from in-memory disk cache anyways.

**Comment by aemoncannon** [comment]
A third alternative would be to write to a temp file then read it in and lie to the compiler about its path: new BatchSourceFile(AbstractFile.getFile(pathToTemp), contents)

**Comment by edani** [comment]
@aemoncannon  Thanks for doing that benchmark. While sending the contents over the wire isn't great, it still isn't horrible (100 requests per second... the fastest typist apparently does 290 wpm, which translates to 24 cps).

**Comment by aemoncannon** [comment]
True. Slower is not necessarily too slow.

**Comment by aemoncannon** [comment]
One contributor to the speed difference might be that reading a string from a sexp stream requires examining every single character. Is this an escape? Is this a delimiter? It's not a great fit for passing a big array of bytes. This could be alleviated if we had a sized vector type in our sexp format. As it stands, I'm leaning toward writing to a tmp file being the best solution (unless there are advantages to passing over the socket that I'm missing?).

**Comment by edani** [comment]
OK, I'll put out a version that uses temp files.
